proxy/upgrade.go

```diff
@@ -0,0 +1,195 @@
+package proxy
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// isUpgradeRequest returns true if and only if there is a "Connection"
+// key with the value "Upgrade" in Headers of the given request.
+func isUpgradeRequest(req *http.Request) bool {
```

Comment:
1) for web sockets, it makes sense to apply the filters for the initial round of establishing the connection. For http2, I don't know yet how to do it, but it would be nice, if we could implement it some time soon. For spdy, as you prefer, here just wanted to make sure that other functionality gets not broken.

2) for web sockets, it is ok, we only care about the initiation phase. For, http2, the same: i don't know yet. SPDY: yes, true, makes not much sense to invest too much efforts to achieve filtering.

3) this depends on the protocol that we upgrade to. based on this: https://github.bus.zalan.do/pathfinder/land-of-mosaic/pull/4, we'll need to care only about 'WebSocket' in the midterm future. At least I hope.

so after all, with the --experimental-upgrade flag, ship it!